### ISSUE TITLE
yolov4tiny quantization issues

### ISSUE THREAD

**Issue by WillLiGitHub** [quic/aimet#1732]
Aimet                                       tf-gpu-1.22.2
AimetCommon                   tf-gpu-1.22.2
AimetTensorflow               tf-gpu-1.22.2

i trained a object detection model by yolov4tiny，mAP@0.5 on dataset is 88%。
but, when i quantized my model with aimet，mAP decreased to 34%.

`
    
    import tensorflow as tf2
    tf = tf2.compat.v1
    tf.disable_eager_execution()
    tf.logging.set_verbosity(tf.logging.ERROR)
    from aimet_common.defs import QuantScheme
    from aimet_tensorflow.common import graph_eval
    from aimet_tensorflow import cross_layer_equalization
    from aimet_tensorflow import bias_correction as aimet_bc
    from aimet_tensorflow import batch_norm_fold as aimet_bnf
    from aimet_tensorflow.quantsim import QuantizationSimModel, save_checkpoint
    from aimet_tensorflow.utils.graph import update_keras_bn_ops_trainable_flag

    
    def add_image_net_computational_nodes_in_graph(session, pred_names, pred_shapes):
      """
      :param session: Tensorflow session to operate on
      :param logits_name: Output tensor name of session graph
      """
      with session.graph.as_default():
          # for n in session.graph.as_graph_def().node:
          #     print(n.name)
          pred_conv2d_12 = session.graph.get_tensor_by_name(pred_names[0])        # (b, 13, 13, 3*(5+num_classes)
          pred_decode_0 = session.graph.get_tensor_by_name(pred_names[1])         # (b, 13*13, 3, (5+num_classes)
          _, gx_0, gy_0, _ = pred_shapes[0]
          _, _, a_0, ch_0 = pred_shapes[1]
          pred_decode_0 = tf.reshape(pred_decode_0,
                                     shape=(-1, pred_conv2d_12.shape[1], pred_conv2d_12.shape[2],
                                            pred_decode_0.shape[2], pred_decode_0.shape[3]))
  
          pred_conv2d_9 = session.graph.get_tensor_by_name(pred_names[2])         # (b, 26, 26, 3*(5+num_classes)
          pred_decode_1 = session.graph.get_tensor_by_name(pred_names[3])         # (b, 26*26, 3, (5+num_classes)
          pred_decode_1 = tf.reshape(pred_decode_1,
                                     shape=(-1, pred_conv2d_9.shape[1], pred_conv2d_9.shape[2],
                                            pred_decode_1.shape[2], pred_decode_1.shape[3]))
  
          _, gx_1, gy_1, _ = pred_shapes[2]
          _, _, a_1, ch_1 = pred_shapes[3]
  
          true_label_0 = tf.placeholder(tf.float32, shape=[None, gx_0, gy_0, a_0, ch_0], name='labels_0')
          true_bboxes_0 = tf.placeholder(tf.float32, shape=[None, YOLO_MAX_BBOX_PER_SCALE, 4], name='bboxes_0')
          true_label_1 = tf.placeholder(tf.float32, shape=[None, gx_1, gy_1, a_1, ch_1], name='labels_1')
          true_bboxes_1 = tf.placeholder(tf.float32, shape=[None, YOLO_MAX_BBOX_PER_SCALE, 4], name='bboxes_1')
  
          giou_loss_0, conf_loss_0, prob_loss_0 = \
              compute_loss(pred_decode_0, pred_conv2d_12, true_label_0, true_bboxes_0, 0, NUM_CLASS=2)
          giou_loss_1, conf_loss_1, prob_loss_1 = \
              compute_loss(pred_decode_1, pred_conv2d_9, true_label_1, true_bboxes_1, 1, NUM_CLASS=2)
  
          giou_loss = giou_loss_0 + giou_loss_1
          conf_loss = conf_loss_0 + conf_loss_1
          prob_loss = prob_loss_0 + prob_loss_1
          loss = giou_loss + conf_loss + prob_loss
          tf.losses.add_loss(giou_loss, 'giou_loss')
          tf.losses.add_loss(conf_loss, 'conf_loss')
          tf.losses.add_loss(prob_loss, 'prob_loss')
          tf.losses.add_loss(loss, tf.GraphKeys.LOSSES)

    def create_quant_sim_model(sess,
                               start_op_names,
                               output_op_names,
                               use_cuda,
                               evaluator,
                               dataset):
        """
        Apply quantizer simulator on the original model and return its object.
        :param sess: The sess with graph.
        :param start_op_names: The list of input op names of the sess.graph
        :param output_op_names: The list of output op names of the sess.graph
        :param use_cuda: If True then use a GPU for QuantizationSimModel
        :param parity_config_file: Config file for H/W parity
        :param evaluator: A callback function that is expected to run forward passes on a session
        :return: QuantizationSimModel object
        """
        quant_scheme = QuantScheme.post_training_tf_enhanced

        rounding_mode = 'nearest'
    
        default_output_bw = 8
    
        default_param_bw = 8
    
        quant_sim_model = QuantizationSimModel(session=sess,
                                               starting_op_names=start_op_names,
                                               output_op_names=output_op_names[0::2],
                                               quant_scheme=quant_scheme, rounding_mode=rounding_mode,
                                               default_output_bw=default_output_bw,
                                               default_param_bw=default_param_bw,
                                               use_cuda=use_cuda)
    
        iterations = 5*dataset.batch_size
    
        print('compute_encodings')
        quant_sim_model.compute_encodings(forward_pass_callback=evaluator,
                                          forward_pass_callback_args=(iterations, dataset, start_op_names, output_op_names[1::2]))
        return quant_sim_model

    model = create_my_yolov4tiny_model(size=320, num_class=2)
    print('load weights: ', TRAIN_CHECKPOINTS_WEIGHTS)
    model.load_weights(TRAIN_CHECKPOINTS_WEIGHTS)

    update_ops_name = [op.name for op in model.updates]
    model = update_keras_bn_ops_trainable_flag(model, trainable=False, load_save_path=TRAIN_CHECKPOINTS_FOLDER)
    sess = tf.keras.backend.get_session()
    output_op_names = [x.name for x in model.outputs]
    print('input names : ', model.input_names)
    print('output names: ', output_op_names)
    add_image_net_computational_nodes_in_graph(sess, output_op_names, model.output_shape)

    # 4. Quantization Sim Model
    # 4.1. Creates Quantization Sim model using AIMET QuantizationSimModel
    print('fold all batch norms')
    applied_sess, _ = aimet_bnf.fold_all_batch_norms(sess,
                                                     input_op_names=[INPUT_NAMES],
                                                     output_op_names=[x.split(":")[0] for x in output_op_names[0::2]])
    print('create quant sim model')
    quant_sim = create_quant_sim_model(sess=applied_sess,
                                       start_op_names=[INPUT_NAMES],
                                       output_op_names=[x.split(":")[0] for x in output_op_names],
                                       use_cuda=True,
                                       evaluator=evaluate,
                                       dataset=testset)

    print("Starting Model QAT")
    # 5.1. Trains the quantization aware model
    train(quant_sim,
          update_ops_name=update_ops_name,
          learning_rate=TRAIN_LR_INIT,
          num_epochs=TRAIN_EPOCHS,
          decay_rate=0.1,
          decay_steps=25,
          debug_steps=300,
          train_dataset=trainset,
          valid_dataset=testset,
          output_names=[x.split(':')[0] for x in output_op_names[1::2]])
`

**Comment by quic-akhobare** [comment]
Hi @WillLiGitHub - we will need to understand the sensitivity of this model to quantization. Some suggested next steps

- After you create the QuantSim model, please run an evaluation so we get a quantized model accuracy score before training
- As another data point, can you change default output and param bw to 16, to see if we get good accuracy for this model with higher integer precision
- Same experiment as above, but we can try param=16, activation=8 and act=16, param = 8. This will tell us if the model is sensitive to weight quantization or activation quant.

**Comment by WillLiGitHub** [comment]
I will give it a try，Thanks a lot!!!

**Comment by ShobithaShivakumar** [comment]
@WillLiGitHub please note that the TensorFlow variant of AIMET has been deprecated. If you are still hitting quantization accuracy issues, we recommend migrating to aimet-torch (latest: 2.26) or aimet-onnx.

 

If you have further questions, feel free to join our http://aihub.qualcomm.com/community/slack specifically the #aimet channel to discuss easily!
